src/Tools/BuildValidator/LocalReferenceResolver.cs

```diff
@@ -0,0 +1,126 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.Extensions.Logging;
+
+namespace BuildValidator
+{
+    /// <summary>
+    /// Resolves references for a package by looking in local nuget and artifact
+    /// directories for Roslyn
+    /// </summary>
+    internal class LocalReferenceResolver : IMetadataReferenceResolver
+    {
+        private readonly Dictionary<Guid, string> _cache = new Dictionary<Guid, string>();
+        private readonly HashSet<DirectoryInfo> _indexDirectories = new HashSet<DirectoryInfo>();
+        private readonly ILogger _logger;
+
+        public LocalReferenceResolver(ILoggerFactory loggerFactory)
+        {
+            _logger = loggerFactory.CreateLogger<LocalReferenceResolver>();
+            _indexDirectories.Add(GetArtifactsDirectory());
+            _indexDirectories.Add(GetNugetCacheDirectory());
+        }
+
+        public static DirectoryInfo GetNugetCacheDirectory()
+        {
+            var nugetPackageDirectory = Environment.GetEnvironmentVariable("NUGET_PACKAGES");
+            if (nugetPackageDirectory is null)
+            {
+                nugetPackageDirectory = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".nuget");
+            }
+
+            return new DirectoryInfo(nugetPackageDirectory);
+        }
+
+        public Task<MetadataReference> ResolveReferenceAsync(MetadataReferenceInfo referenceInfo)
+        {
+            var path = Search(referenceInfo.Mvid, referenceInfo.Name);
+            return Task.FromResult<MetadataReference>(MetadataReference.CreateFromFile(path));
+        }
+
+        public Task<ImmutableArray<MetadataReference>> ResolveReferencesAsync(IEnumerable<MetadataReferenceInfo> references)
+        {
+            var referenceArray = references.ToImmutableArray();
+            CacheNames(referenceArray);
+
+            var files = referenceArray.Select(r => _cache[r.Mvid]);
+
+            var metadataReferences = files.Select(f => MetadataReference.CreateFromFile(f)).Cast<MetadataReference>().ToImmutableArray();
+            return Task.FromResult(metadataReferences);
+        }
+
+        public string Search(Guid mvid, string name)
+        {
+            if (_cache.TryGetValue(mvid, out var referencePath))
+            {
+                return referencePath;
+            }
+
+            foreach (var directory in _indexDirectories)
+            {
+                var foundFile = directory.GetFiles(name, SearchOption.AllDirectories).FirstOrDefault();
+                if (foundFile is null)
+                {
+                    continue;
+                }
+
+                referencePath = foundFile.FullName;
+                _cache[mvid] = referencePath;
```

Comment:
A given file can exist with mulitple MVID in our directory structure. Can't assume that because it has the same name it will have the same MVID. Need to crack the file here and check the MVID.

Reply:
Is that because a file can have multiple modules?  Updating to handle that case

Reply:
Not that. It's that the same DLL can exist in multiple versions within our repo. It's a rare case as we strongly prefer uniform versions but there are cases where we reference multiple versions of same file. Another thing to consider is that when dealing with a NuGet package that has multi-targeted assemblies each assembly for a given TF will have a different MVID but the same name.

Reply:
Ah I see. I was looking at Assembly properties, and it has `Modules` and wrongly assumed that's what you were talking about. Makes sense, and will be handled in next revision